speechbrain/lobes/models/huggingface_transformers/llama.py

```diff
@@ -0,0 +1,133 @@
+
+    Transformer from HuggingFace needs to be installed:
+    https://huggingface.co/transformers/installation.html
+
+    Arguments
+    ---------
+    source : str
+        HuggingFace hub name: e.g "meta-llama/Llama-2-7b-chat-hf"
+    save_path : str
+        Path (dir) of the downloaded model.
+    bnb_config : transformers.BitsAndBytesConfig
+        BitsAndBytesConfig enabling quantisation of the model. If not specified, the model weights will be loaded with weight_precision_load dtype.
+    freeze : bool (default: false)
+        If True, the model is frozen. If False, the model will be trained
+        alongside with the rest of the pipeline.
+    pad_token : str (default: "PAD")
+        String representation of the padding token. This may change from one model to another.
+    padding_side : str (default: "right")
+        One of right or left.
+    weight_precision_load : torch.dtype (default: torch.bfloat16)
+        If no bnb_config is given, this parameter defines the loading type of the parameters of the model. This is useful to reduce memory footprint, but it does not change the compute dtype. For this just refer to mixed precision training in SpeechBrain.
+
+    Example
+    -------
+    >>> model_hub = "meta-llama/Llama-3.2-3B"
+    >>> save_path = "savedir"
+    >>> model = LLaMA(model_hub, save_path)
+    >>> tokens = torch.tensor([[1, 1]])
+    >>> attention_mask = torch.tensor([[1, 1]])
+    >>> outputs = model(tokens, attention_mask)
+    """
+
+    def __init__(
+        self,
+        source: str,
+        save_path: str,
+        bnb_config: BitsAndBytesConfig = None,
+        freeze: bool = False,
+        pad_token: str = "[PAD]",
+        padding_side: Literal["left", "right"] = "right",
+        weight_precision_load: torch.dtype = torch.bfloat16,
+    ) -> None:
+        self.pad_token = pad_token
+        self.padding_side = padding_side
+        self.source = source
+        self.save_path = save_path
+        self.bnb_config = bnb_config
+
+        if self.bnb_config is not None:
+            logger.info(
+                "LlaMA will be quantised following the given configuration."
+            )
+
+        super().__init__(
+            source=source,
+            save_path=save_path,
+            freeze=freeze,
+            with_casual_lm=True,
+            quantization_config=self.bnb_config,
+        )
+
+        if self.bnb_config is None:
+            self.model = self.model.to(weight_precision_load)
+            logger.info(
+                f"Converted {self.__class__.__name__} to {weight_precision_load}"
+            )
+
+        self.load_tokenizer(source=source, pad_token=self.pad_token)
+        self.model.resize_token_embeddings(
+            len(self.tokenizer), pad_to_multiple_of=8
+        )  # TODO: optimal value of pad_to_multiple_of varies? see https://github.com/huggingface/tokenizers/issues/991.
```

Comment:
I believe it should in `huggingface.py`

Reply:
Not all models have embeddings.

Reply:
There's specifics methods in HF to get the input/output embedding of a model. Look at `get_input_embeddings` and `get_output_embeddings`. I think it would be better to move this code in the `huggingface.py` as I believe this is a feature that could be used by other decoder models.

Reply:
@Adel-Moumen I'll check. But I see the fact of putting it in huggingface.py as a problem. The reason is that we are dynamically changing the graph of the model, and this should not be hidden to the user. I feel like, actually, it should stay here but also be much better documented in the docstring. No?